[docs] Verifying provenance with Kyverno

Docs on how to verify provenance generated by the generic workflow with [Kyverno](https://kyverno.io/).

Kyverno docs on verifying images is here: https://kyverno.io/docs/writing-policies/verify-images/

I tested out Kyverno with the generic workflow and it currently doesn't work because of how we store the certificate use to sign the attestations. 

Kyverno seems to call [VerifyImageAttestations](https://github.com/sigstore/cosign/blob/e4bc4a95743bb38922246856d02d987d1272e199/pkg/cosign/verify.go#L587-L608) from the cosign API [here](https://github.com/kyverno/kyverno/blob/fbde1fbcbe6fbe6454c427f5e91ec0e045776f23/pkg/cosign/client.go#L25-L27).

I have the following cluster policy set:
```yaml
apiVersion: kyverno.io/v1
kind: ClusterPolicy
metadata:
  name: check-slsa-attestations
spec:
  validationFailureAction: enforce
  webhookTimeoutSeconds: 30
  rules:
    - name: check-all-keyless
      match:
        any:
          - resources:
              kinds:
                - Pod
      verifyImages:
        - imageReferences:
            # Replace with your image. Wildcard values are supported.
            - "ghcr.io/ianlewis/actions-test:*"
          attestors:
            - entries:
                - keyless:
                    subject: "https://github.com/slsa-framework/slsa-github-generator/.github/workflows/generator_generic_slsa3.yml@refs/tags/v1.1.1"
                    issuer: "https://token.actions.githubusercontent.com"
          attestations:
            - predicateType: https://slsa.dev/provenance/v0.2
              conditions:
                - all:
                    - key: "{{ invocation.configSource.uri }}"
                      operator: Equals
                      value: "git+https://github.com/ianlewis/actions-test@refs/tags/v0.0.11"
                    - key: "{{ invocation.configSource.entryPoint }}"
                      operator: Equals
                      value: ".github/workflows/generic-container.yaml"
                    - key: "{{ regex_match('^https://github.com/slsa-framework/slsa-github-generator/.github/workflows/generator_generic_slsa3.yml@refs/tags/v[0-9].[0-9].[0-9]$','{{ builder.id}}') }}"
                      operator: Equals
                      value: true
```

When I try to create the following Pod:
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: actions-test
spec:
  containers:
    - name: actions-test
      image: ghcr.io/ianlewis/actions-test:v0.0.11
  restartPolicy: Never
```

I get the following error:
```shell
$ kubectl apply -f actions-test-pod.yaml 
Error from server: error when creating "actions-test-pod.yaml": admission webhook "mutate.kyverno.svc-fail" denied the request: 

resource Pod/default/actions-test was blocked due to the following policies

check-slsa-attestations:
  check-all-keyless: |-
    failed to verify signature for ghcr.io/ianlewis/actions-test:v0.0.11: .attestors[0].entries[0].keyless: no matching attestations:
    no certificate found on attestation
```

I'm not sure yet why it's failing but I suspect it has to do with [how we are storing the cert w/ Rekor](https://github.com/slsa-framework/slsa-github-generator/blob/024238bc10cc620be603a7f5329be7a672171ba7/signing/sigstore/rekor.go#L67-L86).

It's the same error you get locally if you try to verify using keyless:
```shell
$ COSIGN_EXPERIMENTAL=1 cosign verify-attestation ghcr.io/ianlewis/actions-test:v0.0.11
Error: no matching attestations:
no certificate found on attestation
main.go:52: error during command execution: no matching attestations:
no certificate found on attestation
```

--- a/internal/builders/container/README.md
+++ b/internal/builders/container/README.md
@@ -22,6 +22,8 @@ project simply generates provenance as a separate step in an existing workflow.
   - [Workflow Inputs](#workflow-inputs)
   - [Provenance Format](#provenance-format)
   - [Provenance Example](#provenance-example)
+- [Verification](#verification)
+  - [Kyverno](#kyverno)
 
 ---
 
@@ -256,4 +258,95 @@ generated as an [in-toto](https://in-toto.io/) statement with a SLSA predicate.
     ]
   }
 }
+```
+
+## Verification
+
+Verification of provenance attestations can be done via several different tools.
+This section shows examples of several popular tools.
+
+## Kyverno
+
+[Kyverno](https://kyverno.io/) is a policy management controller that can be
+used to write Kubernetes-native policies for SLSA provenance. The following
+assumes you have [installed Kyverno](https://kyverno.io/docs/installation/) in
+your Kubernetes cluster.
+
+The following Kyverno `ClusterPolicy` can be used to
+[verify container images](https://kyverno.io/docs/writing-policies/verify-images/)
+as part of [Admission Control](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/).
+
+```yaml
+apiVersion: kyverno.io/v1
+kind: ClusterPolicy
+metadata:
+  name: check-slsa-attestations
+spec:
+  validationFailureAction: enforce
+  webhookTimeoutSeconds: 30
+  rules:
+    - name: check-all-keyless
+      match:
+        any:
+          - resources:
+              kinds:
+                - Pod
+      verifyImages:
+        # imageReferences sets which images the policy will apply to.
+        # Replace with your image. Wildcard values are supported.
+        - imageReferences:
+            - "ghcr.io/ianlewis/actions-test:*"
+          attestors:
+            # This section declares which attestors are accepted. The subject
+            # below corresponds to the OIDC identity of the container workflow.
+            # The issuer corresponds to the GitHub OIDC server that issues the
+            # identity.
+            - entries:
+                - keyless:
+                    subject: "https://github.com/slsa-framework/slsa-github-generator/.github/workflows/generator_container_slsa3.yml@refs/tags/v1.2.0"
+                    issuer: "https://token.actions.githubusercontent.com"
+          # This section declares some policy conditions acting on the provenance itself.
+          attestations:
+            - predicateType: https://slsa.dev/provenance/v0.2
+              conditions:
+                - all:
+                    # This condition verifies that the image was generated from
+                    # the source repository we expect. Replace this with your
+                    # repository.
+                    - key: "{{ invocation.configSource.uri }}"
+                      operator: Equals
+                      value: "git+https://github.com/ianlewis/actions-test@refs/tags/v0.0.11"
+
+                    # This condition verifies the entrypoint of the workflow.
+                    # Replace with the relative path to your workflow in your
+                    #  repository.
+                    - key: "{{ invocation.configSource.entryPoint }}"
+                      operator: Equals
+                      value: ".github/workflows/generic-container.yaml"
+
+                    # This condition verifies that the builder is the builder we
+                    # expect and trust. The following condition can be used
+                    # unmodified. It verifies that the builder is the container
+                    # workflow.
+                    - key: "{{ regex_match('^https://github.com/slsa-framework/slsa-github-generator/.github/workflows/generator_container_slsa3.yml@refs/tags/v[0-9].[0-9].[0-9]$','{{ builder.id}}') }}"
+                      operator: Equals
+                      value: true
+```
+
+When applied the `ClusterPolicy` will be evaluated when a `Pod` is created. If
+the `Pod` fulfills the policy conditions then the `Pod` can be created. If the
+`Pod` does not fulfill one or more of the policy conditions then you will see an
+error like the following. For example, the following error will occur when no
+attestation for the image can be found.
+
+```shell
+$ kubectl apply -f pod.yaml
+Error from server: error when creating "pod.yaml": admission webhook "mutate.kyverno.svc-fail" denied the request:
+
+resource Pod/default/actions-test was blocked due to the following policies
+
+check-slsa-attestations:
+  check-all-keyless: |-
+    failed to verify signature for ghcr.io/ianlewis/actions-test:v0.0.11: .attestors[0].entries[0].keyless: no matching attestations:
+    no certificate found on attestation
 ```

--- a/internal/builders/container/README.md
+++ b/internal/builders/container/README.md
@@ -265,7 +265,7 @@ generated as an [in-toto](https://in-toto.io/) statement with a SLSA predicate.
 Verification of provenance attestations can be done via several different tools.
 This section shows examples of several popular tools.
 
-## Kyverno
+### Kyverno
 
 [Kyverno](https://kyverno.io/) is a policy management controller that can be
 used to write Kubernetes-native policies for SLSA provenance. The following

